runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -390,53 +415,119 @@ public static DataflowRunner fromOptions(PipelineOptions options) {
     return overridesBuilder.build();
   }
 
+  private static class ParDoSingleViaMulti<InputT, OutputT>
+      extends ForwardingPTransform<PCollection<? extends InputT>, PCollection<OutputT>> {
+    private final ParDo.SingleOutput<InputT, OutputT> original;
+
+    public ParDoSingleViaMulti(
+        DataflowRunner ignored, ParDo.SingleOutput<InputT, OutputT> original) {
+      this.original = original;
+    }
+
+    @Override
+    protected PTransform<PCollection<? extends InputT>, PCollection<OutputT>> delegate() {
+      return original;
+    }
+
+    @Override
+    public PCollection<OutputT> expand(PCollection<? extends InputT> input) {
+      TupleTag<OutputT> mainOutput = new TupleTag<>();
+      return input.apply(original.withOutputTags(mainOutput, TupleTagList.empty())).get(mainOutput);
+    }
+  }
+
   private static class ReflectiveOneToOneOverrideFactory<
           InputT, OutputT, TransformT extends PTransform<PCollection<InputT>, PCollection<OutputT>>>
       extends SingleInputOutputOverrideFactory<
           PCollection<InputT>, PCollection<OutputT>, TransformT> {
     private final Class<PTransform<PCollection<InputT>, PCollection<OutputT>>> replacement;
+    @Nullable
     private final DataflowRunner runner;
 
     private ReflectiveOneToOneOverrideFactory(
         Class<PTransform<PCollection<InputT>, PCollection<OutputT>>> replacement,
-        DataflowRunner runner) {
+        @Nullable DataflowRunner runner) {
       this.replacement = replacement;
       this.runner = runner;
     }
 
     @Override
     public PTransformReplacement<PCollection<InputT>, PCollection<OutputT>> getReplacementTransform(
         AppliedPTransform<PCollection<InputT>, PCollection<OutputT>, TransformT> transform) {
+      InstanceBuilder<PTransform<PCollection<InputT>, PCollection<OutputT>>> builder =
+          InstanceBuilder.ofType(replacement);
+      if (runner != null) {
+        builder = builder.withArg(DataflowRunner.class, runner);
+      }
       PTransform<PCollection<InputT>, PCollection<OutputT>> rep =
-          InstanceBuilder.ofType(replacement)
-              .withArg(DataflowRunner.class, runner)
+          builder
               .withArg(
                   (Class<TransformT>) transform.getTransform().getClass(), transform.getTransform())
               .build();
       return PTransformReplacement.of(PTransformReplacements.getSingletonMainInput(transform), rep);
     }
   }
 
+  private static class ReflectiveCollectionToTupleOverrideFactory<
```

Comment:
I'm very much not a fan of the Reflective override factories, and in fact would prefer to be moving away from them. Is the burden really too high to add two separate factories that do something obvious?